### PR TITLE
[Sprint: 50] XD-2717 Add nameExpression and directoryExpression properties to File Sink

### PR DESCRIPTION
**Pull request by fmarchand**
I have signed and agree to the terms of the SpringSource Individual
Contributor License Agreement.

### CODE REVIEW

**Comment by fmarchand** [comment]
I tried something using this PR.
Input payload is a json containing a field `date` :

   `... | file --nameExpression='#jsonPath(payload,''$.date'')'"`

Spring XD tells me that method jsonPath is null ... 

if I try :

`... | log --expression='#jsonPath(payload,''$.date'')'"`

it works ...

I don't see why ! Could someone help me to fix that. Otherwise this PR is incomplete I think.

**Comment by fmarchand** [comment]
I'm adding dirExpression to this module don't merge now please I will have finished in 30 minutes ...

**Comment by markpollack** [comment]
Can one option cover the full path as in the hdfs sink?

**Comment by fmarchand** [comment]
I'm gonna check the hdfs sink to see if I can do that. It's a good idea ! That would be easier to use.

**Comment by fmarchand** [comment]
I don't see the option you're talking about in HDFS sink ... I see `directory` but it's not a SpEL expression.

**Comment by markpollack** [comment]
Look at http://docs.spring.io/spring-xd/docs/current/reference/html/#hadoop-hdfs and --partitionPath

**Comment by markpollack** [comment]
Maybe it is in fact two distinct options after all...

**Comment by fmarchand** [comment]
Oh .... I see ... because I use the same options from Spring Integration `file:outbound-adapter` I  kept the two distinct options. I did this pull request because I need it in one of my projects and keeping the two options is easier in my case and the stream is more readable when you display it with `stream list`. Don't you think ?

**Comment by markpollack** [comment]
Yea, you are right.  Mapping to the underlying options in SI is better.  Sorry for the noise

**Comment by fmarchand** [comment]
No pb :) That's always a pleasure to talk with you pivotal guys :) I always learn a lot from you. I'm gonna finish my PR. Hope it will be ok.

**Comment by fmarchand** [comment]
Job done.

**Comment by ghillert** [comment]
@fmarchand Thanks for the PR!! Could you change `DirExpression` to `DirectoryExpression` (properties, the module options etc.). I think this will be more consistent with the underlying Spring Integration.

**Comment by fmarchand** [comment]
I've done the 'directory' thing

**Comment by ghillert** [comment]
@fmarchand Sorry for my earlier confusion. I merged the PR. Only at second looks I realized that we were already using `dir`. In that context dirExpression actually makes sense. Thanks for the contribution!!!!

**Comment by fmarchand** [comment]
You're welcome ! NP for the confusion. That happens :)

File: spring-xd-shell/src/test/java/org/springframework/xd/shell/command/FileSourceAndFileSinkTests.java [hunk 1]
```diff
@@ -113,4 +120,38 @@ private void testFileSinkSource(String sinkSuffix, String sinkParam) throws Exce
...
+		try {
+			FileCopyUtils.copy("hello", new FileWriter(in));
+			String directoryExpression = " --directoryExpression='''/tmp/'' + ''" + dirPrefix + "'' + payload.trim() + ''"
+					+ dirSuffix + "'''";
+			stream().create(streamName, "in: file --outputType=text/plain | out: file " + nameExpressionParam +
+					directoryExpression);
+			Thread.sleep(1000);
+			String actual = FileCopyUtils.copyToString(new FileReader(out));
+			assertEquals("hello", actual.trim());
+			assertEquals(out.exists(), is(true));
```
Comment (garyrussell) [inline spring-xd-shell/src/test/java/org/springframework/xd/shell/command/FileSourceAndFileSinkTests.java]: @ghillert I am getting a failure here on master; this should be `assertThat()` (or simply `assertTrue`).

`true` can never equal the hamcrest matcher `Is`.

File: spring-xd-shell/src/test/java/org/springframework/xd/shell/command/FileSourceAndFileSinkTests.java [hunk 2]
```diff
@@ -113,4 +120,38 @@ private void testFileSinkSource(String sinkSuffix, String sinkParam) throws Exce
...
+		File outDir = new File("/tmp/" + dirPrefix + "hello" + dirSuffix);
+		File out = new File(outDir, "hello");
+
+		try {
+			FileCopyUtils.copy("hello", new FileWriter(in));
+			String directoryExpression = " --directoryExpression='''/tmp/'' + ''" + dirPrefix + "'' + payload.trim() + ''"
+					+ dirSuffix + "'''";
+			stream().create(streamName, "in: file --outputType=text/plain | out: file " + nameExpressionParam +
+					directoryExpression);
+			Thread.sleep(1000);
```
Comment (garyrussell) [inline spring-xd-shell/src/test/java/org/springframework/xd/shell/command/FileSourceAndFileSinkTests.java]: Also, arbitrary sleeps like this are unreliable on the CI servers and can extend the overall test times; it's better to intelligently wait for some condition (e.g. loop 100 times for 100ms until the condition is true).

It provides a fast exit when the condition is generally true faster than the arbitrary sleep, while allowing for a longer time when running on a heavily loaded server.

File: spring-xd-shell/src/test/java/org/springframework/xd/shell/command/FileSourceAndFileSinkTests.java (hunk 1 above)
Comment (dturanski) [inline spring-xd-shell/src/test/java/org/springframework/xd/shell/command/FileSourceAndFileSinkTests.java]: This test should use the same pattern as others in here, i.e., https://github.com/spring-projects/spring-xd/blob/9356b584b1765304e16a1c2db5450e28aebab38b/spring-xd-shell/src/test/java/org/springframework/xd/shell/command/FileSourceAndFileSinkTests.java#L87

Comment (fmarchand) [inline spring-xd-shell/src/test/java/org/springframework/xd/shell/command/FileSourceAndFileSinkTests.java]: @garyrussel : yes it should be assertThat ... and that's what I'm used to use ... my mistake ! But what bothers me the most : why I did not get exception when I ran the gradlew dist ? I missed something ...

**Pull request closed without merging**